From e9caedb9740e8dcf3b50e44b976fcb0124fedcc3 Mon Sep 17 00:00:00 2001
From: Catalin Udma <catalin-dan.udma@nxp.com>
Date: Tue, 29 Sep 2020 13:06:09 +0300
Subject: [PATCH] xen: extend xen_swiotlb_dma_ops with map_resource and
 unmap_resource

dma_map_resource requires a map_resource callback to be non-NULL,
otherwise it returns a mapping error.
fsl-edma DMA driver uses dma_map_resource for 5.4, added from upstream
in commit:
	0fa89f972da60: dmaengine: fsl-edma: dma map slave device address

When using fsl-edma over Xen, dma_map_resource requires the map_resource
callback to be implemented in xen_swiotlb_dma_ops.

Add map_resource and unmap_resources for xen_swiotlb using direct
mapping similar to dma_direct_map_resource()

Issue: ALB-5458
Signed-off-by: Catalin Udma <catalin-dan.udma@nxp.com>
Signed-off-by: Andrei Cherechesu <andrei.cherechesu@nxp.com>
---
 drivers/xen/swiotlb-xen.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/drivers/xen/swiotlb-xen.c b/drivers/xen/swiotlb-xen.c
index bd3a10d..21a2160 100644
--- a/drivers/xen/swiotlb-xen.c
+++ b/drivers/xen/swiotlb-xen.c
@@ -563,6 +563,19 @@ xen_swiotlb_dma_supported(struct device *hwdev, u64 mask)
 	return xen_virt_to_bus(xen_io_tlb_end - 1) <= mask;
 }
 
+static dma_addr_t xen_swiotlb_map_resource(struct device *dev, phys_addr_t phys,
+		size_t size, enum dma_data_direction dir, unsigned long attrs)
+{
+	dma_addr_t dma_addr = phys;
+
+	return dma_addr;
+}
+
+static void xen_swiotlb_unmap_resource(struct device *dev, dma_addr_t handle,
+		size_t size, enum dma_data_direction dir, unsigned long attrs)
+{
+}
+
 const struct dma_map_ops xen_swiotlb_dma_ops = {
 	.alloc = xen_swiotlb_alloc_coherent,
 	.free = xen_swiotlb_free_coherent,
@@ -575,6 +588,8 @@ const struct dma_map_ops xen_swiotlb_dma_ops = {
 	.map_page = xen_swiotlb_map_page,
 	.unmap_page = xen_swiotlb_unmap_page,
 	.dma_supported = xen_swiotlb_dma_supported,
+	.map_resource = xen_swiotlb_map_resource,
+	.unmap_resource = xen_swiotlb_unmap_resource,
 	.mmap = dma_common_mmap,
 	.get_sgtable = dma_common_get_sgtable,
 };
-- 
2.7.4

